gather.go

```diff
@@ -146,46 +147,62 @@ func (a *Agent) gatherCandidates() {
 		a.log.Warnf("Failed to run onCandidateHdlr task: %v\n", err)
 		return
 	}
-	a.gatheringState = GatheringStateComplete
 
+	if err := a.run(func(agent *Agent) {
+		a.gatheringState = GatheringStateComplete
+	}); err != nil {
+		a.log.Warnf("Failed to update gatheringState: %v\n", err)
+		return
+	}
 }
 
 func (a *Agent) gatherCandidatesLocal(networkTypes []NetworkType) {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
 	localIPs := localInterfaces(networkTypes)
+	wg.Add(len(localIPs) * len(supportedNetworks))
 	for _, ip := range localIPs {
 		for _, network := range supportedNetworks {
-			conn, err := listenUDP(int(a.portmax), int(a.portmin), network, &net.UDPAddr{IP: ip, Port: 0})
-			if err != nil {
-				a.log.Warnf("could not listen %s %s\n", network, ip)
-				continue
-			}
+			go func(network string, ip net.IP) {
+				conn, err := listenUDP(int(a.portmax), int(a.portmin), network, &net.UDPAddr{IP: ip, Port: 0})
+				if err != nil {
+					a.log.Warnf("could not listen %s %s\n", network, ip)
+					wg.Done()
+					return
+				}
 
-			port := conn.LocalAddr().(*net.UDPAddr).Port
-			c, err := NewCandidateHost(network, ip, port, ComponentRTP)
-			if err != nil {
-				a.log.Warnf("Failed to create host candidate: %s %s %d: %v\n", network, ip, port, err)
-				continue
-			}
+				port := conn.LocalAddr().(*net.UDPAddr).Port
+				c, err := NewCandidateHost(network, ip, port, ComponentRTP)
+				if err != nil {
+					a.log.Warnf("Failed to create host candidate: %s %s %d: %v\n", network, ip, port, err)
+					wg.Done()
+					return
+				}
 
-			if err := a.run(func(agent *Agent) {
-				set := a.localCandidates[c.NetworkType()]
-				set = append(set, c)
-				a.localCandidates[c.NetworkType()] = set
-			}); err != nil {
-				a.log.Warnf("Failed to append to localCandidates: %v\n", err)
-				continue
-			}
+				if err := a.run(func(agent *Agent) {
```

Comment:
this is async, wg.Done can be called before the candidate is added, i suggest another wg (or maybe sync.Condition) here

Reply:
sync.Cond is likely to go away in go 2. I would recommend against that.
https://github.com/golang/go/issues/21165

Reply:
p.s. `a.run` is not really async. It uses an unbuffered channel for the tasks which means only one task is running at a time. In this case, every `a.run` waits for it's turn to write into the channel. Only after that point it runs separately. What happens if you have another task running that is stalling?! Kinda challenging design with this `a.run` but it works.

Reply:
@masterada If you still feel strongly about this specific comment let me know. I will try to figure out how to address it.

Reply:
nvm, you set `a.gatheringState = GatheringStateComplete` in a.run as well, so it definitely runs after all candidates are gathered (sorry, was looking on it on mobile :) )